skyportal/facility_apis/lt.py

```diff
@@ -0,0 +1,855 @@
+                "format": "date",
+                "title": "End Date (UT)",
+                "default": (datetime.utcnow().date() + timedelta(days=7)).isoformat(),
+            },
+            "maximum_airmass": {
+                "title": "Maximum Airmass (1-3)",
+                "type": "number",
+                "default": 2.0,
+            },
+            "maximum_seeing": {
+                "title": "Maximum Seeing [arcsec] (0-5)",
+                "type": "number",
+                "default": 1.2,
+            },
+            "sky_brightness": {
+                "title": "Maximum allowable Sky Brightness, Dark + X magnitudes [arcsec] (0-5)",
+                "type": "number",
+                "default": 2.0,
+            },
+            "photometric": {
+                "title": "Does this observation require photometric conditions?",
+                "type": "boolean",
+            },
+            "binning": {"type": "string", "enum": ["1x1", "2x2"], "default": "1x1"},
+        },
+        "required": [
+            "observation_type",
+            "exposure_type",
+            "priority",
+            "start_date",
+            "end_date",
+            "maximum_airmass",
+            "maximum_seeing",
+            "binning",
+        ],
+    }
+
+    ui_json_schema = {}
+
+
+class IOIAPI(LTAPI):
+
+    """An interface to LT IOI operations."""
+
+    # subclasses *must* implement the method below
+    @staticmethod
+    def submit(request):
+
+        """Submit a follow-up request to LT's IOI.
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to add to the queue and the SkyPortal database.
+        """
+
+        from ..models import DBSession, FacilityTransaction
+
+        altdata = request.allocation.load_altdata()
+        if not altdata:
+            return
+
+        ltreq = IOIRequest()
+        observation_payload = ltreq._build_prolog()
+        ltreq._build_project(observation_payload, request)
+        ltreq._build_inst_schedule(observation_payload, request)
+
+        headers = {
+            'Username': altdata["username"],
+            'Password': altdata["password"],
+        }
+        url = '{0}://{1}:{2}/node_agent2/node_agent?wsdl'.format(
+            'http', cfg['app.lt_host'], cfg['app.lt_port']
+        )
+        client = Client(url=url, headers=headers)
+        full_payload = etree.tostring(
+            observation_payload, encoding="unicode", pretty_print=True
+        )
+        # Send payload, and receive response string, removing the encoding tag which causes issue with lxml parsing
+        response = client.service.handle_rtml(full_payload).replace(
+            'encoding="ISO-8859-1"', ''
+        )
+        response_rtml = etree.fromstring(response)
+        mode = response_rtml.get('mode')
+
+        if mode == 'confirm':
+            request.status = 'submitted'
+        else:
+            request.status = f'rejected: {mode}'
```

Comment:
can we display the reason the request was rejected

Reply:
Can you help me understand how to do this?

Reply:
`response_rtml` looks like

```
<?xml version="1.0"  standalone="no"?>
<RTML xmlns="http://www.rtml.org/v3.1a" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" mode="reject" uid="1605337885" version="3.1a" xsi:schemaLocation="http://www.rtml.org/v3.1a http://telescope.livjm.ac.uk/rtml/RTML-nightly.xsd">
  <History>
    <Entry timeStamp="2020-11-14T07:11:33+00:00">
      <Agent name="unknown" uri="unknown"/>
      <Description>java.rmi.ServerException: RemoteException occurred in server thread; nested exception is:
	java.rmi.RemoteException: Exception while handling request: java.rmi.ServerException: RemoteException occurred in server thread; nested exception is:
	java.rmi.RemoteException: com.mysql.jdbc.MysqlDataTruncation; nested exception is:
	com.mysql.jdbc.MysqlDataTruncation: Data truncation: Data truncated for column 'name' at row 1</Description>
      <Error>java.rmi.ServerException: RemoteException occurred in server thread; nested exception is:
	java.rmi.RemoteException: Exception while handling request: java.rmi.ServerException: RemoteException occurred in server thread; nested exception is:
	java.rmi.RemoteException: com.mysql.jdbc.MysqlDataTruncation; nested exception is:
	com.mysql.jdbc.MysqlDataTruncation: Data truncation: Data truncated for column 'name' at row 1</Error>
    </Entry>
    <Entry timeStamp="2020-11-14T07:11:33+00:00">
      <Agent name="LT_NodeAgent" uri="urn:/node_agent"/>
      <Description>RemoteException occurred in server thread; nested exception is:
	java.rmi.RemoteException: Exception while handling request: java.rmi.ServerException: RemoteException occurred in server thread; nested exception is:
	java.rmi.RemoteException: com.mysql.jdbc.MysqlDataTruncation; nested exception is:
	com.mysql.jdbc.MysqlDataTruncation: Data truncation: Data truncated for column 'name' at row 1</Description>
      <Error>java.rmi.ServerException: RemoteException occurred in server thread; nested exception is:
	java.rmi.RemoteException: Exception while handling request: java.rmi.ServerException: RemoteException occurred in server thread; nested exception is:
	java.rmi.RemoteException: com.mysql.jdbc.MysqlDataTruncation; nested exception is:
	com.mysql.jdbc.MysqlDataTruncation: Data truncation: Data truncated for column 'name' at row 1</Error>
    </Entry>
  </History>
  <Project ProjectID="GrowthTest">
    <Contact>
      <Username>fritz_bot</Username>
      <Name>fritz_bot</Name>
      <Communication/>
    </Contact>
  </Project>
  <Schedule>
    <Target name="cef9a353-6ccc-4aae-9fe0-e77173186741">
      <Coordinates>
        <RightAscension>
          <Hours>0</Hours>
          <Minutes>0</Minutes>
          <Seconds>0.0</Seconds>
          <Offset units="arcseconds">0.0</Offset>
        </RightAscension>
        <Declination>
          <Degrees>-0</Degrees>
          <Arcminutes>0</Arcminutes>
          <Arcseconds>0.0</Arcseconds>
          <Offset units="arcseconds">0.0</Offset>
        </Declination>
        <Equinox>J2000</Equinox>
      </Coordinates>
    </Target>
    <Device name="IO:O" type="camera">
      <SpectralRegion>optical</SpectralRegion>
      <Setup>
        <Filter type="G"/>
        <Detector>
          <Binning>
            <X units="pixels">1</X>
            <Y units="pixels">1</Y>
          </Binning>
        </Detector>
      </Setup>
    </Device>
    <Priority>3</Priority>
    <Exposure count="2">
      <Value units="seconds">150.0</Value>
    </Exposure>
    <DateTimeConstraint type="include">
      <DateTimeStart system="UT" value="2020-11-14T00:00:00+00:00"/>
      <DateTimeEnd system="UT" value="2020-11-21T00:00:00+00:00"/>
    </DateTimeConstraint>
    <SeeingConstraint maximum="1.2" minimum="0.0" units="arcseconds"/>
    <SkyConstraint>
      <Flux>2.0</Flux>
      <Units>magnitudes/square-arcsecond</Units>
    </SkyConstraint>
    <ExtinctionConstraint>
      <Clouds>clear</Clouds>
    </ExtinctionConstraint>
    <AirmassConstraint maximum="2.0" minimum="0.0"/>
  </Schedule>
  <Schedule>
    <Target name="cef9a353-6ccc-4aae-9fe0-e77173186741">
      <Coordinates>
        <RightAscension>
          <Hours>0</Hours>
          <Minutes>0</Minutes>
          <Seconds>0.0</Seconds>
          <Offset units="arcseconds">0.0</Offset>
        </RightAscension>
        <Declination>
          <Degrees>-0</Degrees>
          <Arcminutes>0</Arcminutes>
          <Arcseconds>0.0</Arcseconds>
          <Offset units="arcseconds">0.0</Offset>
        </Declination>
        <Equinox>J2000</Equinox>
      </Coordinates>
    </Target>
    <Device name="IO:O" type="camera">
      <SpectralRegion>optical</SpectralRegion>
      <Setup>
        <Filter type="R"/>
        <Detector>
          <Binning>
            <X units="pixels">1</X>
            <Y units="pixels">1</Y>
          </Binning>
        </Detector>
      </Setup>
    </Device>
    <Priority>3</Priority>
    <Exposure count="2">
      <Value units="seconds">150.0</Value>
    </Exposure>
    <DateTimeConstraint type="include">
      <DateTimeStart system="UT" value="2020-11-14T00:00:00+00:00"/>
      <DateTimeEnd system="UT" value="2020-11-21T00:00:00+00:00"/>
    </DateTimeConstraint>
    <SeeingConstraint maximum="1.2" minimum="0.0" units="arcseconds"/>
    <SkyConstraint>
      <Flux>2.0</Flux>
      <Units>magnitudes/square-arcsecond</Units>
    </SkyConstraint>
    <ExtinctionConstraint>
      <Clouds>clear</Clouds>
    </ExtinctionConstraint>
    <AirmassConstraint maximum="2.0" minimum="0.0"/>
  </Schedule>
  <Schedule>
    <Target name="cef9a353-6ccc-4aae-9fe0-e77173186741">
      <Coordinates>
        <RightAscension>
          <Hours>0</Hours>
          <Minutes>0</Minutes>
          <Seconds>0.0</Seconds>
          <Offset units="arcseconds">0.0</Offset>
        </RightAscension>
        <Declination>
          <Degrees>-0</Degrees>
          <Arcminutes>0</Arcminutes>
          <Arcseconds>0.0</Arcseconds>
          <Offset units="arcseconds">0.0</Offset>
        </Declination>
        <Equinox>J2000</Equinox>
      </Coordinates>
    </Target>
    <Device name="IO:O" type="camera">
      <SpectralRegion>optical</SpectralRegion>
      <Setup>
        <Filter type="I"/>
        <Detector>
          <Binning>
            <X units="pixels">1</X>
            <Y units="pixels">1</Y>
          </Binning>
        </Detector>
      </Setup>
    </Device>
    <Priority>3</Priority>
    <Exposure count="2">
      <Value units="seconds">150.0</Value>
    </Exposure>
    <DateTimeConstraint type="include">
      <DateTimeStart system="UT" value="2020-11-14T00:00:00+00:00"/>
      <DateTimeEnd system="UT" value="2020-11-21T00:00:00+00:00"/>
    </DateTimeConstraint>
    <SeeingConstraint maximum="1.2" minimum="0.0" units="arcseconds"/>
    <SkyConstraint>
      <Flux>2.0</Flux>
      <Units>magnitudes/square-arcsecond</Units>
    </SkyConstraint>
    <ExtinctionConstraint>
      <Clouds>clear</Clouds>
    </ExtinctionConstraint>
    <AirmassConstraint maximum="2.0" minimum="0.0"/>
  </Schedule>
</RTML>
```

So I think we just need to extract the content of the `<Error>` tag.